acir/src/circuit/opcodes/black_box_function_call.rs

```diff
@@ -7,138 +7,376 @@ use serde::{Deserialize, Serialize};
-            write_u32(&mut writer, input.num_bits)?;
+    let mut inputs = Vec::with_capacity(num_inputs as usize);
+
+    for _ in 0..num_inputs {
+        let witness_index = read_u32(&mut reader)?;
+        inputs.push(Witness::new(witness_index));
+    }
+
+    Ok(inputs)
+}
+
+impl BlackBoxFuncCall {
+    pub fn get_black_box_func(&self) -> BlackBoxFunc {
+        match self {
+            BlackBoxFuncCall::AES { .. } => BlackBoxFunc::AES,
+            BlackBoxFuncCall::AND { .. } => BlackBoxFunc::AND,
+            BlackBoxFuncCall::XOR { .. } => BlackBoxFunc::XOR,
+            BlackBoxFuncCall::RANGE { .. } => BlackBoxFunc::RANGE,
+            BlackBoxFuncCall::SHA256 { .. } => BlackBoxFunc::SHA256,
+            BlackBoxFuncCall::Blake2s { .. } => BlackBoxFunc::Blake2s,
+            BlackBoxFuncCall::ComputeMerkleRoot { .. } => BlackBoxFunc::ComputeMerkleRoot,
+            BlackBoxFuncCall::SchnorrVerify { .. } => BlackBoxFunc::SchnorrVerify,
+            BlackBoxFuncCall::Pedersen { .. } => BlackBoxFunc::Pedersen,
+            BlackBoxFuncCall::HashToField128Security { .. } => BlackBoxFunc::HashToField128Security,
+            BlackBoxFuncCall::EcdsaSecp256k1 { .. } => BlackBoxFunc::EcdsaSecp256k1,
+            BlackBoxFuncCall::FixedBaseScalarMul { .. } => BlackBoxFunc::FixedBaseScalarMul,
+            BlackBoxFuncCall::Keccak256 { .. } => BlackBoxFunc::Keccak256,
         }
+    }
 
-        let num_outputs = self.outputs.len() as u32;
-        write_u32(&mut writer, num_outputs)?;
+    pub fn get_inputs_vec(&self) -> Vec<FunctionInput> {
+        match self {
+            BlackBoxFuncCall::AES { inputs, .. }
+            | BlackBoxFuncCall::SHA256 { inputs, .. }
+            | BlackBoxFuncCall::Blake2s { inputs, .. }
+            | BlackBoxFuncCall::Keccak256 { inputs, .. }
+            | BlackBoxFuncCall::Pedersen { inputs, .. }
+            | BlackBoxFuncCall::HashToField128Security { inputs, .. } => inputs.to_vec(),
+            BlackBoxFuncCall::AND { lhs, rhs, .. } | BlackBoxFuncCall::XOR { lhs, rhs, .. } => {
+                vec![*lhs, *rhs]
+            }
+            BlackBoxFuncCall::FixedBaseScalarMul { input, .. }
+            | BlackBoxFuncCall::RANGE { input } => vec![*input],
+            BlackBoxFuncCall::ComputeMerkleRoot { leaf, index, hash_path, .. } => {
+                let mut inputs = Vec::new();
+                inputs.push(*leaf);
+                inputs.push(*index);
+                inputs.extend(hash_path.iter().cloned());
+                inputs
+            }
+            BlackBoxFuncCall::SchnorrVerify {
+                public_key_x,
+                public_key_y,
+                signature,
+                message,
+                ..
+            } => {
+                let mut inputs = Vec::new();
+                inputs.push(*public_key_x);
+                inputs.push(*public_key_y);
+                inputs.extend(signature.iter().cloned());
```

Comment:
```suggestion
                inputs.extend(signature.iter().copied());
```
Can you change the other instances of `.cloned()` to `.copied()`. This signals that we're doing something cheap here.